server/src/test/java/org/elasticsearch/cluster/metadata/DataStreamTests.java

```diff
@@ -78,4 +78,23 @@ public void testRollover() {
         assertTrue(rolledDs.getIndices().containsAll(ds.getIndices()));
         assertTrue(rolledDs.getIndices().contains(newWriteIndex));
     }
+
+    public void testRemoveBackingIndex() {
+        int numBackingIndices = randomIntBetween(2, 32);
+        int indexToRemove = randomIntBetween(1, numBackingIndices - 1);
+        String dataStreamName = randomAlphaOfLength(10).toLowerCase(Locale.ROOT);
+
+        List<Index> indices = new ArrayList<>(numBackingIndices);
+        for (int k = 1; k <= numBackingIndices; k++) {
+            indices.add(new Index(DataStream.getBackingIndexName(dataStreamName, k), UUIDs.randomBase64UUID(random())));
+        }
+        DataStream original = new DataStream(dataStreamName, "@timestamp", indices);
+        DataStream updated = original.removeBackingIndex(indices.get(indexToRemove - 1));
+        assertThat(updated.getName(), equalTo(original.getName()));
+        assertThat(updated.getGeneration(), equalTo(original.getGeneration()));
+        assertThat(updated.getTimeStampField(), equalTo(original.getTimeStampField()));
+        for (int k = 0; k < (numBackingIndices - 1); k++) {
+            assertThat(updated.getIndices().get(k), equalTo(original.getIndices().get(k < (indexToRemove - 1) ? k : k + 1)));
+        }
```

Comment:
maybe also assert that the index that is supposed to be removed really is removed? I think we now just check that all other indices but the removed index exists. I think that for that, just checking that the number of backing indices is decremented by 1 should suffice.